Add blog post about sampling with tool calls (#1934)
modelcontextprotocol/modelcontextprotocol: blog/content/posts/2025-12-xx-sampling-supports-tool-calls.md

+++
date = '2025-12-12T12:00:00Z'
title = 'Closing the Loop: Tool Support Comes to MCP Sampling'
author = 'Jonathan Hefner'
tags = ['sampling', 'tools', 'agentic']
+++

Tool use transformed LLMs from sophisticated text generators into agents capable of taking action in the world. Before tool use, you could ask an LLM about the weather and get a plausible-sounding guess. With tool use, it can actually check.

MCP extended that power to the ecosystem level — a standard way for agents to discover and invoke tools across any number of servers. But there was an asymmetry hiding in the architecture. While LLMs could be agentic, the tools they called could not. Even tools that used MCP's sampling feature to request LLM completions were limited to simple, one-shot text generation.

> [SEP-1577](https://github.com/modelcontextprotocol/modelcontextprotocol/issues/1577) has changed that. MCP sampling now supports tool calling, which means tools themselves can drive agentic workflows.

Reviewer: Would it make sense to link to a markdown version of this SEP instead?

MCP extended that power to the ecosystem level, becoming the standard way for agents to discover and invoke tools across any number of servers. But there was an asymmetry hiding in the architecture. While LLMs could be agentic, the tools they called could not. Even tools that used MCP's [sampling feature](https://modelcontextprotocol.io/specification/2025-11-25/client/sampling) — which lets servers request completions from the host application's LLM — were limited to simple, one-shot text generation.

[SEP-1577](https://github.com/modelcontextprotocol/modelcontextprotocol/issues/1577), now part of the MCP specification, has changed that. Sampling supports tool calling, which means tools themselves can drive agentic workflows.

## The asymmetry in the architecture

Today's tool calls follow a simple pattern: an LLM reasons, invokes a tool, gets a result, and continues reasoning. This works well when tools are simple functions. But what if a tool needs to be smart enough to reason, make decisions, or coordinate multiple steps?

> MCP's sampling feature was designed for exactly this: letting tools request LLM completions. But sampling only supported simple text generation. A server could make multiple sampling calls and chain them together with code, but that code couldn't reason. It could follow predetermined logic paths, but it couldn't adapt on the fly. The server could ask the LLM questions, but the LLM couldn't drive the investigation. This effectively capped sampling's power at pre-tool-calling levels — tools could get answers from an LLM, but they couldn't delegate agentic behavior to one.

Reviewer 2: ```suggestion
MCP's sampling feature was designed for exactly this scenario: letting tools request LLM completions. But until now, sampling only supported simple text generation. A server could make multiple sampling calls and chain them together with code, but that code couldn't reason - it could only follow predetermined logic. The server could ask the LLM questions, but the LLM couldn't take the lead and drive the investigation itself. This limited sampling to pre-tool-calling capabilities: tools could get answers from an LLM, but they couldn't hand off agentic behavior to one.
```

```typescript
const result = await mcpServer.server.createMessage({
  messages: [
    {
      role: "user",
      content: {
        type: "text",
        text: "Research the latest developments in fusion energy and summarize the key breakthroughs.",
      },
    },
  ],
  tools: [
    {
      name: "web_search",
      description: "Search the web for current information",
      inputSchema: {
        type: "object",
        properties: {
          query: { type: "string", description: "Search query" },
        },
        required: ["query"],
      },
    },
  ],
  toolChoice: { mode: "auto" },
  maxTokens: 2000,
});
```

When the LLM decides to use a tool, the response comes back with `stopReason: "toolUse"` and includes `tool_use` content blocks describing what the LLM wants to invoke. The server then executes those tools, sends the results back in a follow-up sampling request, and the LLM continues, potentially calling more tools until it's ready to produce a final response.

> This is the standard agentic loop, but now it can happen _inside_ a tool implementation.

Reviewer 2: ```suggestion
This is known as the **standard agentic loop**, but now it can happen _inside_ a tool implementation.
```

6. Your server executes the tool and gets a result.
7. Your server calls `sampling/createMessage` again, including the tool result.
8. Steps 4-7 repeat until the LLM returns `stopReason: "endTurn"`.
9. Your tool returns its result to the original LLM.
10. That LLM can now call more tools, each potentially agentic themselves.

```mermaid
sequenceDiagram
    participant Host as Host Application / LLM
    participant Server as Server / Tool
    participant Sub as Server / Sub-tools

    Host->>Server: tools/call("research")
    activate Server
    Note right of Server: Tool needs LLM reasoning

    Server->>Host: sampling/createMessage<br/>{ messages, tools }

    loop
        Note left of Host: LLM reasons,<br/>decides to call tool
        Host-->>Server: { stopReason: "toolUse",<br/>content: [tool_use] }
        Server->>Sub: Execute sub-tool
        Sub-->>Server: Sub-tool result
        Server->>Host: sampling/createMessage<br/>{ messages: [..., tool_result], tools }
    end

    Host-->>Server: { stopReason: "endTurn",<br/>content: [text] }
    Note right of Server: Tool uses LLM response in result
    Server-->>Host: tools/call result
    deactivate Server
```

> The key insight is that the server drives the tool loop. The host application's job is to provide LLM access and maintain human oversight. The server decides what tools to expose, executes them when called, and determines when the loop is complete.

Reviewer 2: ```suggestion
The server drives the tool loop. The host application's job is simply to provide LLM access and maintain human oversight. The server decides which tools to offer, runs them when called, and determines when the process is complete.
```

> Here's a generic implementation of this pattern that you can adapt for your own tools:

Reviewer 2: ```suggestion
Here's a generic implementation of this pattern that you can adapt for your own tools with the [MCP TypeScript SDK](https://github.com/modelcontextprotocol/typescript-sdk):
```

**Data analysis tools** can explore datasets, form hypotheses, run queries to test them, and produce insights, not just execute predetermined queries.

**Document processing tools** can read documents, extract information, and produce structured outputs, handling complexity internally through iterative reasoning.

The pattern extends naturally. Any tool that would benefit from reasoning, iteration, or multi-step workflows can now implement them directly.

## Capability negotiation

Not all clients will support tool use in sampling immediately. SEP-1577 adds a [`sampling.tools`](https://modelcontextprotocol.io/specification/2025-11-25/schema#clientcapabilities-sampling) capability flag so servers can detect support.

Servers should check for this capability flag at connection time. For tools that require agentic sampling to function, the simplest approach is to omit the tool from `tools/list` when the capability is absent. For tools that can function with degraded capability, consider implementing a graceful fallback, perhaps noting the limitation in the tool's output so the outer LLM has context.

## A note on `includeContext`

> SEP-1577 also soft-deprecates the `"thisServer"` and `"allServers"` values of the `includeContext` parameter. These values — which allowed servers to request that the client include conversation context from the current or all connected servers — had ambiguous semantics that made it difficult for clients to implement consistently. This ambiguity contributed to low adoption of sampling overall.

Reviewer 2: ```suggestion
SEP-1577 also soft-deprecates the `"thisServer"` and `"allServers"` values of the `includeContext` parameter. These values allowed servers to request conversation context from the current server or all connected servers, but their meaning was ambiguous and made it hard for clients to implement them consistently. This ambiguity contributed to low adoption of sampling overall.
```

## Getting started

SEP-1577 is part of MCP version `2025-11-25`. See the [sampling specification](https://modelcontextprotocol.io/specification/2025-11-25/client/sampling) for full protocol details.

SDK support is available in:

- **TypeScript SDK**: Version 1.23.0+ ([PR #1101](https://github.com/modelcontextprotocol/typescript-sdk/pull/1101))
- **Python SDK**: Version 1.23.0+ ([PR #1594](https://github.com/modelcontextprotocol/python-sdk/pull/1594))

> To experiment with agentic sampling, update to an SDK version that includes these changes and ensure you're connecting to a client that advertises the `sampling.tools` capability.

Author: No, there are no clients that support this today.  That is the primary reason is PR is marked as draft.  However, it might be worth publishing anyway, to encourage adoption.

> To experiment with agentic sampling, update to an SDK version that includes these changes and ensure you're connecting to a client that advertises the `sampling.tools` capability.

Reviewer 3: I do plan to get that sorted and get the feature in this month tho! :)
